src/shell/braces.zig

```diff
@@ -0,0 +1,805 @@
+            const group_alloc = group_alloc_.get();
+            var group = std.ArrayList(AST.Atom).init(group_alloc);
+            var close = false;
+            while (!self.match(.eof)) {
+                if (self.match(.close)) {
+                    close = true;
+                    break;
+                }
+                if (self.match(.comma)) break;
+                const group_atom = try self.parseAtom() orelse break;
+                try group.append(group_atom);
+            }
+            if (group.items.len == 1) {
+                try variants.append(.{ .atoms = .{ .single = group.items[0] } });
+            } else {
+                try variants.append(.{ .atoms = .{ .many = group.items[0..] } });
+            }
+            if (close) break;
+        }
+
+        return .{ .variants = variants.items[0..] };
+    }
+
+    fn has_eq_sign(self: *Parser, str: []const u8) ?u32 {
+        _ = self;
+        return @import("./shell.zig").hasEqSign(str);
+    }
+
+    fn advance(self: *Parser) Token {
+        if (!self.is_at_end()) {
+            self.current += 1;
+        }
+        return self.prev();
+    }
+
+    fn is_at_end(self: *Parser) bool {
+        return self.peek() == .eof;
+    }
+
+    fn expect(self: *Parser, toktag: TokenTag) Token {
+        std.debug.assert(toktag == @as(TokenTag, self.peek()));
+        if (self.check(toktag)) {
+            return self.advance();
+        }
+        unreachable;
+    }
+
+    /// Consumes token if it matches
+    fn match(self: *Parser, toktag: TokenTag) bool {
+        if (@as(TokenTag, self.peek()) == toktag) {
+            _ = self.advance();
+            return true;
+        }
+        return false;
+    }
+
+    fn match_any2(self: *Parser, comptime toktags: []const TokenTag) ?Token {
+        const peeked = self.peek();
+        inline for (toktags) |tag| {
+            if (peeked == tag) {
+                _ = self.advance();
+                return peeked;
+            }
+        }
+        return null;
+    }
+
+    fn match_any(self: *Parser, comptime toktags: []const TokenTag) bool {
+        const peeked = @as(TokenTag, self.peek());
+        inline for (toktags) |tag| {
+            if (peeked == tag) {
+                _ = self.advance();
+                return true;
+            }
+        }
+        return false;
+    }
+
+    fn check(self: *Parser, toktag: TokenTag) bool {
+        return @as(TokenTag, self.peek()) == @as(TokenTag, toktag);
+    }
+
+    fn peek(self: *Parser) Token {
+        return self.tokens[self.current];
+    }
+
+    fn peek_n(self: *Parser, n: u32) Token {
+        if (self.current + n >= self.tokens.len) {
+            return self.tokens[self.tokens.len - 1];
+        }
+
+        return self.tokens[self.current + n];
+    }
+
+    fn prev(self: *Parser) Token {
+        return self.tokens[self.current - 1];
+    }
+
+    fn add_error(self: *Parser, comptime fmt: []const u8, args: anytype) !void {
```

Comment:
we're not totally consistent about this, but generally in Zig we should do camelCase function names and snake_cased field names